clientapi/routing/register_test.go

```diff
@@ -264,3 +279,318 @@ func TestSessionCleanUp(t *testing.T) {
 		}
 	})
 }
+
+func Test_validateUsername(t *testing.T) {
+	tooLongUsername := strings.Repeat("a", maxUsernameLength)
+	tests := []struct {
+		name      string
+		localpart string
+		domain    gomatrixserverlib.ServerName
+		want      *util.JSONResponse
+	}{
+		{
+			name:      "empty username",
+			localpart: "",
+			domain:    "localhost",
+			want: &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.InvalidUsername("Username can only contain characters a-z, 0-9, or '_-./='"),
+			},
+		},
+		{
+			name:      "invalid username",
+			localpart: "INVALIDUSERNAME",
+			domain:    "localhost",
+			want: &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.InvalidUsername("Username can only contain characters a-z, 0-9, or '_-./='"),
+			},
+		},
+		{
+			name:      "username too long",
+			localpart: tooLongUsername,
+			domain:    "localhost",
+			want: &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.BadJSON(fmt.Sprintf("%q exceeds the maximum length of %d characters", fmt.Sprintf("@%s:%s", tooLongUsername, "localhost"), maxUsernameLength)),
+			},
+		},
+		{
+			name:      "localpart starting with an underscore",
+			localpart: "_notvalid",
+			domain:    "localhost",
+			want: &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.InvalidUsername("Username cannot start with a '_'"),
+			},
+		},
+		{
+			name:      "valid username",
+			localpart: "valid",
+			domain:    "localhost",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := validateUsername(tt.localpart, tt.domain); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("validateUsername() = %v, want %v", got, tt.want)
+			}
+			if got := validateApplicationServiceUsername(tt.localpart, tt.domain); !reflect.DeepEqual(got, tt.want) {
+				if got != nil && got.JSON != jsonerror.InvalidUsername("Username cannot start with a '_'") {
+					t.Errorf("validateUsername() = %v, want %v", got, tt.want)
+				}
+			}
+		})
+	}
+}
+
+func Test_validatePassword(t *testing.T) {
+	tests := []struct {
+		name     string
+		password string
+		want     *util.JSONResponse
+	}{
+		{
+			name:     "password too short",
+			password: "shortpw",
+			want: &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: min %d chars", minPasswordLength)),
```

Comment:
I generally dislike doing direct checks on error strings, as they aren't really part of the public API. We should be checking the `errcode`. Write a helper function and `gjson` out the code? This applies throughout these tests.

Reply:
Like above, we should just remove the JSON here and actually return error types. (Same for the other `validate*` functions)